backend/composer/services/cs_ingestion/cs_ingestion_services.py

```diff
@@ -40,253 +39,351 @@
-        for entity in statement[VIAS]:
-            if not_found_entity(entity[ENTITY_URI]):
-                invalid_entities.append({"entity": entity[ENTITY_URI], "prop": VIAS})
-
-    if len(invalid_entities) > 0:
-        logging.warning(
-            f'Skip statement {statement[LABEL]} due to the following entities not found in composer db: {invalid_entities}')
-        return True
-    return False
-
-
-def has_multiple_destination(statement):
-    if len(statement[DESTINATION]) > 1:
-        logging.warning(f'Skip statement {statement[LABEL]} due to multiple destinations')
-        return True
-    else:
-        return False
-
-
-def has_invalid_sex(statement):
-    if has_prop(statement[SEX]) and len(Sex.objects.filter(ontology_uri=statement[SEX][0])) == 0:
-        logging.warning(f'Skip statement {statement[LABEL]} due to sex {statement[SEX][0]} not found in composer db')
-        return True
-    else:
-        return False
-
-
-def has_invalid_species(statement):
-    if has_prop(statement[SPECIES]):
-        for s in statement[SPECIES]:
-            try:
-                Specie.objects.get(ontology_uri=s)
-                return False
-            except:
-                logging.warning(f'Skip statement {statement[LABEL]} due to specie {s} not found in composer db')
-                return True
-    return False
+def ingest_statements():
+    statements_list = get_statements_from_neurondm()
 
+    valid_statements = validate_statements(statements_list)
+    for statement in valid_statements:
+        sentence, sentence_created = get_or_create_sentence(statement)
 
-def get_valid_phenotypes(statement):
-    valid_phenotypes = []
-    if has_prop(statement[PHENOTYPE]):
-        for p in statement[PHENOTYPE]:
-            try:
-                phenotype = Phenotype.objects.get(ontology_uri=p)
-                valid_phenotypes.append(phenotype)
-            except:
-                logging.warning(f'Skip phenotype {p} at statement {statement[LABEL]} not found in composer db ')
-    return valid_phenotypes
+        connectivity_statement, statement_created = get_or_create_connectivity_statement(statement, sentence)
 
+        if not statement_created:
+            if should_overwrite(connectivity_statement, statement):
+                defaults = generate_connectivity_statement_defaults(statement, sentence)
+                ConnectivityStatement.objects.filter(id=connectivity_statement.id).update(**defaults)
+                update_many_to_many_fields(connectivity_statement, statement)
+                add_ingestion_system_note(connectivity_statement)
 
-def pick_first_phenotype(phenotypes_list, statement):
-    if len(phenotypes_list) == 0:
-        return None
-    elif len(phenotypes_list) > 1:
-        logging.warning(
-            f'Multiple phenotypes found, ignore the following {phenotypes_list[1:]} at statement {statement[LABEL]}')
-    return phenotypes_list[0]
+    update_forward_connections(valid_statements)
```

Comment:
We do this after all the statements are ingested because there could be references in the forward connection to statements to be ingested in the current batch

Reply:
question: @afonsobspinto what will happen in case the ingestion of a specific statement (let's call it StatA) was rejected and another statement (StatB) has a forward connection to StatA

imo StatB should also not be ingested because of data lost

May be the whole ingestion is an atomic operation: only do it if everything can be processed

cc: @ddelpiano

Reply:
I addressed the first part of the question [here](https://github.com/MetaCell/sckan-composer/pull/202#discussion_r1450993973).

I also wrapped the whole ingestion in a transaction